### DISCUSSION TITLE
Route conditions

### DISCUSSION THREAD

**Discussion by oprypkhantc** [laravel/framework#43438] [Ideas]
Hey.

Symfony has a feature in their router called conditions. It basically allows specifying additional conditions for a route to match through their proprietary language. This comes in handy when dealing with dynamic routes - e.g. those that depend on the host, for example. Laravel's routing is limited to regular expressions, but sometimes you need more than that, like find a user in the DB from a subdomain and route based on that.

I'm proposing adding a `->condition(callable $condition)` method on `Route`. Callable accepts `Request` and returns `true` if route matches or `false` if router should find a different route (or a fallback route, or display a 404). 

This plays nicely with both uncached and cached routes. The former just need a new [`Validator`](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Routing/Matching/ValidatorInterface.php) and the latter can provide a `condition` callback to Symfony's URI matcher by setting index `4` here:
https://github.com/laravel/framework/blob/9.x/src/Illuminate/Routing/RouteCollection.php#L262
https://github.com/symfony/routing/blob/6.1/Matcher/Dumper/CompiledUrlMatcherDumper.php#L126

I'd love to implement this, but I don't want to waste time if it's declined, hence asking here.
@driesvints Any chance for this feature to be accepted?

**Comment by henzeb** [comment]
I'd like to have that logic in a more cleaner place. Like in a controller or an invokable class which you could pass by FQCN, just like you can do with the controller class. 

You'd still call that condition method inside your routes file, but the logic for that condition is placed elsewhere. This makes your routes file a lot cleaner and prevents spaghetti code...

If laravel will not accept such a feature, something like this could easily be a package, as Route is macroable.

**Reply by oprypkhantc** [reply]
Unfortunately having this in a controller simply will not work, as this suggests a given route has to first be matched. The idea is that a route is not matched (e.g. we don't know which controller will handle given request) until conditions are met.

Regarding the package this will also not work :( For it to be production-ready it needs to work properly with `route:cache` and for that we'll have to make modifications to `CompiledRouteCollection`.

**Reply by henzeb** [reply]
True, but my request for accepting FCQN as a parameter still stands.
